maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -123,107 +135,177 @@ public fun GoogleMap(
+                        log("View attached!")
+                        lifecycleOwner = mapView.findViewTreeLifecycleOwner()!!.also {
+                            it.lifecycle.addObserver(lifecycleObserver)
+                        }
+                    }
+
+                    override fun onViewDetachedFromWindow(v: View) {
+                        log("View detached!")
+                        unregisterLifecycleObserver()
+                        lifecycleObserver.moveToBaseState()
+                    }
+                }
+
+                mapView.addOnAttachStateChangeListener(onAttachStateListener)
             }
+        },
+        onReset = { /* View is detached. */ },
+        onRelease = { mapView ->
+            mapView.log("onRelease")
+
+            mapView.tagData().let { tagData ->
+                tagData.componentCallbacks?.let { componentCallbacks ->
+                    tagData.mapViewContext?.unregisterComponentCallbacks(componentCallbacks)
+                }
+
+                tagData.lifecycleObserver!!.moveToDestroyedState()
+            }
+
+            mapView.tag = null
+        },
+        update = { mapView ->
+            // Create Composition
+            if (!isCompositionSet) {
+                isCompositionSet = true
+                debugMapId = mapView.tagData().debugId
+                mapUpdaterScope.launchComposition(mapView)
+            }
+        }
+    )
+
+    Box(modifier = modifier) {
+        Column(
+            modifier = Modifier
+                .align(Alignment.TopStart)
+                .padding(8.dp)
+        ) {
+            Text("Map reused: $debugIsMapReused")
+            Text("Composition ID: $debugCompositionId")
+            Text("MapView ID: $debugMapId")
         }
     }
 }
 
-internal suspend inline fun disposingComposition(factory: () -> Composition) {
-    val composition = factory()
-    try {
-        awaitCancellation()
-    } finally {
-        composition.dispose()
+private fun MapView.registerAndSaveNewComponentCallbacks(context: Context) {
+    val newComponentCallbacks = this.componentCallbacks()
+    val tagData = tagData()
+    tagData.componentCallbacks = newComponentCallbacks
+    tagData.mapViewContext = context
+    context.registerComponentCallbacks(newComponentCallbacks)
+}
+
+internal data class MapTagData(
+    var componentCallbacks: ComponentCallbacks?,
+    var mapViewContext: Context?,
+    var lifecycleObserver: MapLifecycleEventObserver?,
+    val debugId: Int = nextId
+) {
+    companion object {
+        private var nextId = 0
+            get() = field++
+    }
+}
+
+// TODO make private
+internal fun MapView.tagData(): MapTagData = (tag as? MapTagData) ?: let { mapView ->
+    MapTagData(null, null, null).also { newTag ->
+        mapView.tag = newTag
     }
 }
 
-private suspend inline fun MapView.newComposition(
-    parent: CompositionContext,
+private suspend fun MapView.createComposition(
```

Comment:
What do you think about inlining this one?

Reply:
Yes, makes sense 👍